Gem/Code/Source/GameLift/GameLiftServerComponent.cpp

```diff
@@ -0,0 +1,211 @@
+                      "MultiplayerSampleGameLiftServer", "System Component for the GameLift dedicated server")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("System"))
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true);
+            }
+        }
+    }
+
+    void GameLiftServerSystemComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        provided.push_back(AZ_CRC_CE("MultiplayerSampleGameLiftServerService"));
+    }
+
+    void GameLiftServerSystemComponent::GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible)
+    {
+        incompatible.push_back(AZ_CRC_CE("MultiplayerSampleGameLiftServerService"));
+    }
+
+    void GameLiftServerSystemComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("NetworkingService"));
+        required.push_back(AZ_CRC_CE("MultiplayerService"));
+        required.push_back(AZ_CRC_CE("AWSGameLiftServerService"));
+    }
+
+    void GameLiftServerSystemComponent::GetDependentServices(AZ::ComponentDescriptor::DependencyArrayType& dependent)
+    {
+        AZ_UNUSED(dependent);
+    }
+
+    void GameLiftServerSystemComponent::Init()
+    {
+        m_backfillProcessTerminated = true;
+        m_backfillComplete = true;
+    }
+
+    void GameLiftServerSystemComponent::Activate()
+    {
+        AzFramework::SessionNotificationBus::Handler::BusConnect();
+
+        AWSGameLift::AWSGameLiftServerRequestBus::Broadcast(&AWSGameLift::AWSGameLiftServerRequestBus::Events::NotifyGameLiftProcessReady);
+    }
+
+    void GameLiftServerSystemComponent::Deactivate()
+    {
+        StopBackfillProcess();
+
+        AzFramework::SessionNotificationBus::Handler::BusDisconnect();
+    }
+
+    bool GameLiftServerSystemComponent::OnSessionHealthCheck()
+    {
+        return true;
+    }
+
+    bool GameLiftServerSystemComponent::OnCreateSessionBegin(const AzFramework::SessionConfig& sessionConfig)
+    {
+        m_maxPlayer = sessionConfig.m_maxPlayer;
+
+        return true;
+    }
+
+    void GameLiftServerSystemComponent::OnCreateSessionEnd()
+    {
+        StartBackfillProcess();
+    }
+
+    bool GameLiftServerSystemComponent::OnDestroySessionBegin()
+    {
+        StopBackfillProcess();
+
+        return true;
+    }
+
+    void GameLiftServerSystemComponent::OnDestroySessionEnd()
+    {
+        m_maxPlayer = 0;
+    }
+
+    void GameLiftServerSystemComponent::OnUpdateSessionBegin(
+        const AzFramework::SessionConfig& sessionConfig, const AZStd::string& updateReason)
+    {
+        AZ_UNUSED(sessionConfig);
+        if (updateReason == "BACKFILL_FAILED" || updateReason == "BACKFILL_TIMED_OUT" || updateReason == "BACKFILL_CANCELLED")
+        {
+            m_backfillComplete = true;
+            m_waitEvent.release();
```

Comment:
Where is this released on success? I'm not sure what the callback is for completed backfill

Reply:
The idea here is to unblock the acquirer when backfill fails so that the side thread will retry the backfill immediately. I think we can just wait for the 30s time period to pass (m_waitEvent.try_acquire_for(AZStd::chrono::seconds(30))) on success and start another backfill request if there're empty slots.

If it's possible to subscribe to the IMultiplayer interface and react to slot availability, then we may not need the semaphore at all. I'll look into the interface.

Reply:
To be clear, if the IMultiplayer interface isn't meeting that need you should feel free to update it so that it does. Locking mechanisms are generally costly so eliminating that plus the wait/spin would be positive.